app/controllers/application_controller/timelines.rb

```diff
@@ -102,42 +102,54 @@ def tl_build_timeline_report_options
       end
 
       temp_clause = @tl_record.event_where_clause(@tl_options.evt_type)
+      conditions = [temp_clause[0], "timestamp >= ?", "timestamp <= ?"]
+      parameters = temp_clause[1..-1] + [from_dt, to_dt]
 
-      cond = "( "
-      cond = cond << temp_clause[0]
-      params = temp_clause.slice(1, temp_clause.length)
-
-      event_set = @tl_options.event_set
-      if !event_set.empty?
-        if @tl_options.policy_events? && @tl_options.policy.result != "both"
-          where_clause = [") and (timestamp >= ? and timestamp <= ?) and (event_type in (?)) and (result = ?)",
-                          from_dt,
-                          to_dt,
-                          event_set.flatten,
-                          @tl_options.policy.result]
-        else
-          where_clause = [") and (timestamp >= ? and timestamp <= ?) and (event_type in (?))",
-                          from_dt,
-                          to_dt,
-                          event_set.flatten]
-        end
-      else
-        where_clause = [") and (timestamp >= ? and timestamp <= ?)",
-                        from_dt,
-                        to_dt]
-      end
-      cond << where_clause[0]
+      tl_add_event_type_conditions(conditions, parameters)
+      tl_add_policy_conditions(conditions, parameters) if @tl_options.policy_events?
 
-      params2 = where_clause.slice(1, where_clause.length - 1)
-      params = params.concat(params2)
-      @report.where_clause = [cond, *params]
+      condition = "(" + conditions.reduce { |a, c| "#{a}) and (#{c}" } + ")"
```

Comment:
I *really* don't like the SQL building...there has to be a way to leverage ActiveRecord relation chains.

Reply:
Though I do recognize that the original code also did some level of building SQL as well, which is also not good.

Reply:
I will dig deeper and determine what report does with that SQL where clause.

Reply:
I looked and report is just an MiqReport instance... MiqReport#where_clause is a string column that must get used later.  I'm fine with this for now, considering your PR gets the SQL building down to basically just 2 lines.  I still think there;s gotta be a way to build it using ActveRecord though :)

Reply:
Getting closer... 😄 

```ruby
query = @report.db.constantize.where("timestamp >= ?", from_dt).where("timestamp <= ?", to_dt)
query.arel.where_sql
# => "WHERE (timestamp >= '2018-08-05 15:55:43.491733') AND (timestamp <= '2018-08-10 15:55:43.492248')"
```